### PR TITLE
drivers: regulator: enable support for regulator modes

### PR DESCRIPTION
**Pull request by danieldegrasse**
Enable support for regulator modes. Some regulators, such as the PCA9420, offer support for several modes, with each mode containing a discrete set of voltage levels. This PR adds support for managing regulator modes to the `regulator_pmic` driver, and enables it for the PCA9420.

### CODE REVIEW

**Review by galak — COMMENTED** [review]
Feels like we should possibly have proper regulator nodes?  Since these concepts seem to be supported by the regulator binding (if I'm understanding things correctly).

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/regulator/regulator.yaml#n94

**Comment by danieldegrasse** [comment]
@galak by proper regulator nodes, do you mean the regulator framework Linux utilizes? If I'm understanding it correctly, the regulator framework in Linux seems to allow users to get references to regulators based on their name, so the usage pattern a user would follow would be to:
1. get reference to a `struct device` implementing the regulator API
2. get reference to underlying regulator (of which there may be multiple) on the attached device, using the regulator name as a string

IMO, the implementation of the second step feels like it will add unnecessary overhead, and feels like a re implementation of `device_get_binding`

I think that there is probably a broader discussion to be had about expanding the regulator framework/APIs to better support PMICs, but I'm not sure if the way to do it is to reuse the Linux regulator implementation.

**Comment by galak** [comment]
> 

@danieldegrasse I'm more talking about how we describing the info in the devicetree.  We don't need a full regulator API.

**Comment by danieldegrasse** [comment]
> @danieldegrasse I'm more talking about how we describing the info in the devicetree. We don't need a full regulator API.

Ah, makes sense. I'll update this PR to better align with Linux bindings

**Comment by danieldegrasse** [comment]
@galak I've updated the bindings for modes to align with Linux. Unfortunately, this means that the PCA9420 can no longer be handled by the generic regulator driver. Since there are no additional users of the pmic regulator driver, and since the PMIC API is already implemented as a vendor specific API, I have refactored the `regulator_pmic` driver and binding to refer to `nxp,regulator-pca9420`

File: boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts [hunk 1]
```diff
@@ -170,6 +172,22 @@ arduino_serial: &flexcomm12 {
 			enable-reg = <PCA9420_MODECFG_0_2>;
 			enable-mask = <PCA9420_MODECFG_2_SW1_EN_MASK>;
 			enable-val = <PCA9420_MODECFG_2_SW1_EN_VAL>;
+			/*
+			 * This mode-reg setting will permite mode control by
```
Comment (mmahadevan108) [inline boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:176]: permit

File: boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts [hunk 2]
```diff
@@ -170,6 +172,22 @@ arduino_serial: &flexcomm12 {
...
+			 * pins, a feature of the PCA9240. If mode control via
+			 * I2C is desired, the following settings
+			 * should be used:
+			 * regulator-initial-mode = <PCA9420_MODECFG0_I2C>;
+			 * regulator-allowed-modes = <PCA9420_MODECFG0_I2C>,
+			 *			<PCA9420_MODECFG1_I2C>,
+			 *			<PCA9420_MODECFG2_I2C>,
+			 *			<PCA9420_MODECFG3_I2C>;
+			 */
+			regulator-initial-mode = <PCA9420_MODECFG0_I2C>;
```
Comment (mmahadevan108) [inline boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:186]: Do we need to set these new properties for each voltage output?

**Comment by danieldegrasse** [comment]
@mmahadevan108 I've reworked this PR to keep support for regulators generic with the regulator_pmic driver, while still using linux compatible bindings for `regulator-allowed-modes`. I have also updated the devicetree binding for PMIC regulators so that the `regulator-pmic` compatible is added to the PMIC IC itself, allowing properties like `regulator-allowed-modes` to be shared by all regulator outputs

File: include/zephyr/dt-bindings/regulator/pca9420_i2c.h [hunk 3]
```diff
@@ -118,6 +120,7 @@
 	1850000 0x60		/* LDO1 output voltage 1.850V. */\
 	1875000 0x70		/* LDO1 output voltage 1.875V. */\
 	1900000 0x80		/* LDO1 output voltage 1.900V. */\
+	1900000 0x81		/* LDO1 output voltage 1.900V. */\
```
Comment (mmahadevan108) [inline include/zephyr/dt-bindings/regulator/pca9420_i2c.h:123]: This should be` 0x90`

File: include/zephyr/dt-bindings/regulator/pca9420_i2c.h [hunk 4]
```diff
@@ -145,6 +148,7 @@
 	2050000 0x16		/* LDO2 output voltage 2.050V. */\
 	2075000 0x17		/* LDO2 output voltage 2.075V. */\
 	2100000 0x18		/* LDO2 output voltage 2.100V. */\
+	2100000 0x19		/* LDO2 output voltage 2.100V. */\
```
Comment (mmahadevan108) [inline include/zephyr/dt-bindings/regulator/pca9420_i2c.h:151]: Should we be setting `1A-1F` to `2.1V`

File: boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts [hunk 5]
```diff
@@ -156,9 +156,28 @@ arduino_serial: &flexcomm12 {
...
+		 *			<PCA9420_MODECFG3_I2C>;
+		 * modesel-reg = <PCA9420_TOP_CNTL3>;
+		 * modesel-mask = <PCA9420_TOP_CNTL3_MODE_I2C_MASK>;
+		 */
+		regulator-allowed-modes = <PCA9420_MODECFG0_PIN>,
+					<PCA9420_MODECFG1_PIN>,
+					<PCA9420_MODECFG2_PIN>,
+					<PCA9420_MODECFG3_PIN>;
+		regulator-initial-mode = <PCA9420_MODECFG0_PIN>;
+		modesel-reg = <PCA9420_MODECFG_0_0>;
```
Comment (mmahadevan108) [inline boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:178]: Should there be a `modesel-reg` per mode

File: drivers/regulator/regulator_pmic.c [hunk 6]
```diff
@@ -236,6 +278,51 @@ int regulator_get_current_limit(const struct device *dev)
...
+	} else {
+		/* Select mode via I2C */
+		/* Clear register offset */
+		data->reg_offset = 0;
+		rc = regulator_modify_register(dev, config->modesel_reg,
+			mode & PMIC_I2C_MODE_MASK, config->modesel_mask);
+		if (rc) {
+			return rc;
+		}
+		data->reg_offset = sel_off;
```
Comment (mmahadevan108) [inline drivers/regulator/regulator_pmic.c:322]: Why is this line needed?

File: include/zephyr/dt-bindings/regulator/pca9420_i2c.h (hunk 3 above)
Comment (danieldegrasse) [reply include/zephyr/dt-bindings/regulator/pca9420_i2c.h]: Fixed in latest push

File: include/zephyr/dt-bindings/regulator/pca9420_i2c.h (hunk 4 above)
Comment (danieldegrasse) [reply include/zephyr/dt-bindings/regulator/pca9420_i2c.h]: We could- my rationale was that we should assume this driver is the only consumer of the regulator, so these values will never be written (although they are valid)

File: boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts (hunk 5 above)
Comment (danieldegrasse) [reply boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:178]: The `modesel-reg` will be offset based on which mode is selected, so we only need to encode the first `modesel-reg` value in DTS

File: drivers/regulator/regulator_pmic.c (hunk 6 above)
Comment (danieldegrasse) [reply drivers/regulator/regulator_pmic.c:322]: For I2C mode selection (without offset being applied to `modesel_reg`), we first clear the register offset, and make the mode selection. Then, the mode selection offset is applied. For offset based selection (pin mode on the PCA9420), we apply the offset first, since we want it to be applied to `modesel_reg`

**Review by mmahadevan108 — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
